### PR TITLE
Support configuring loadBalancerIP in envoy svc

### PR DESCRIPTION
**Pull request by arkodg**
Fixes: https://github.com/envoyproxy/gateway/issues/1841

### CODE REVIEW

File: api/v1alpha1/shared_types.go [hunk 1]
```diff
@@ -191,6 +191,13 @@ type KubernetesServiceSpec struct {
 	// +optional
 	AllocateLoadBalancerNodePorts *bool `json:"allocateLoadBalancerNodePorts,omitempty"`
 
+	// LoadBalancerIP defines the IP Address of the underlying load balancer service. This field
+	// may be ignored if the load balancer provider does not support this feature.
+	// This field has been Deprecated, but it is still used for set the IP Address in some cloud
+	// providers such as GCP.
```
Comment (zirain) [inline api/v1alpha1/shared_types.go:197]: it's deprecated, maybe use annotations to support this?

Comment (arkodg) [reply api/v1alpha1/shared_types.go:197]: Not possible for GCP, the linked issue has all the details:)

Comment (zirain) [reply api/v1alpha1/shared_types.go:197]: thanks for clearify.

Comment (zirain) [reply api/v1alpha1/shared_types.go:197]: what will happen if someone config this by mistake on non-GCP?

Comment (arkodg) [reply api/v1alpha1/shared_types.go:197]: https://github.com/envoyproxy/gateway/issues/1841#issuecomment-1699472763
has the details, looks like all still support it except AWS and I'm unsure what happens when it is set, any idea @LanceEa ?

Comment (LanceEa) [reply api/v1alpha1/shared_types.go:197]: @zirain - see discussion here: https://github.com/envoyproxy/gateway/issues/1841#issuecomment-1699472763

tl;dr; using term "deprecated" is probably not correct but rather discouraged until k8s community/cloud providers have completely moved away from it.

Comment (zirain) [reply api/v1alpha1/shared_types.go:197]: Just want to make sure, it won't break anything.

Comment (arkodg) [reply api/v1alpha1/shared_types.go:197]: Thanks @LanceEa can you please review the API doc string in this PR

Comment (LanceEa) [reply api/v1alpha1/shared_types.go:197]: My guess is either it is ignored or like GCP the admission controller would reject the fields it doesn't support.

File: api/v1alpha1/shared_types.go [hunk 2]
```diff
@@ -191,6 +191,13 @@ type KubernetesServiceSpec struct {
 	// +optional
 	AllocateLoadBalancerNodePorts *bool `json:"allocateLoadBalancerNodePorts,omitempty"`
 
+	// LoadBalancerIP defines the IP Address of the underlying load balancer service. This field
+	// may be ignored if the load balancer provider does not support this feature.
+	// This field has been Deprecated, but it is still used for set the IP Address in some cloud
```
Comment (LanceEa) [inline api/v1alpha1/shared_types.go:196]: nit: `set` --> `setting`

Might also be good to clarify that it is deprecated in k8s so maybe `This field has been Deprecated in K8s`. 

If you wanted to get fancy you could add a link but since this gets bundled into the CRD definition, this seems sufficient. Also, I think we have good history in the PR that users can search if have questions.

File: api/v1alpha1/shared_types.go (hunk 1 above)
Comment (zirain) [reply api/v1alpha1/shared_types.go:197]: What I'm thinking is should EG auto-detect cloud provider, and reject/ignore this field if it's non-GCP?

File: api/v1alpha1/shared_types.go (hunk 2 above)
Comment (arkodg) [reply api/v1alpha1/shared_types.go:196]: thanks, addressed

File: api/v1alpha1/shared_types.go (hunk 1 above)
Comment (LanceEa) [reply api/v1alpha1/shared_types.go:197]: @arkodg - I agree. At that point I think we would be over engineering  it and would end up with more pain maintaining it. Instead I think leaning on documentation is a better approach.

**Review by zirain — DISMISSED** [review]

File: api/v1alpha1/validation/validate.go [hunk 3]
```diff
@@ -95,6 +96,15 @@ func validateService(spec *egv1a1.EnvoyProxySpec) []error {
 				errs = append(errs, fmt.Errorf("allocateLoadBalancerNodePorts can only be set for %v type", egv1a1.ServiceTypeLoadBalancer))
 			}
 		}
+		if serviceType, serviceLoadBalancerIP := spec.Provider.Kubernetes.EnvoyService.Type, spec.Provider.Kubernetes.EnvoyService.LoadBalancerIP; serviceType != nil && serviceLoadBalancerIP != nil {
+			if *serviceType != egv1a1.ServiceTypeLoadBalancer {
+				errs = append(errs, fmt.Errorf("loadBalancerIP can only be set for %v type", egv1a1.ServiceTypeLoadBalancer))
+			}
+
+			if net.ParseIP(*serviceLoadBalancerIP) == nil {
```
Comment (LanceEa) [inline api/v1alpha1/validation/validate.go:104]: Not a blocker for me but ParseIP will also parse an IPV6 address successfully so it could be nice to give the user earlier feedback.

But realistically most users setting this field have read the Kubernetes docs or can at least search them easily.

**Review by LanceEa — APPROVED** [review]

File: api/v1alpha1/validation/validate.go (hunk 3 above)
Comment (arkodg) [reply api/v1alpha1/validation/validate.go]: thanks for flagging this, adding a check to only allow ipv4

**Review by AliceProxy — APPROVED** [review]
